h2/utilities.py

```diff
@@ -242,3 +249,41 @@ def _reject_pseudo_header_fields(headers):
             seen_regular_header = True
 
         yield header
+
+
+def _check_host_authority_header(headers, connection_state):
+    """
+    Raises a ProtocolError if a header block arrives that does not contain
+    :authority or a Host header, or if a header block contains both fields,
+    but their values do not match.
+    """
+    # We only expect to see :authority and Host headers on request header
+    # blocks that aren't trailers, so skip this validation if we're on the
+    # server side or looking at trailer blocks.
+    if connection_state.is_client or connection_state.is_trailer:
+        for header in headers:
+            yield header
+        return
+
+    authority_header_val = None
+    host_header_val = None
+
+    for header in headers:
+        if header[0] == b':authority':
+            authority_header_val = header[1]
+        elif header[0] == b'host':
+            host_header_val = header[1]
+
+        yield header
+
+    if (authority_header_val is None) and (host_header_val is None):
+        raise ProtocolError(
+            "Did not receive an :authority or Host header."
+        )
+
+    if (authority_header_val is not None) and (host_header_val is not None):
```

Comment:
Oh, that's quite right. In that case, rather than have the nested `if` blocks (which have already confused me once!), let's have sequential `if` blocks. That lets us more easily have a comment that explains what the `None` checks are.